### PR TITLE
Attaching dangling References to models and tests for cloning models

### LINKED ISSUE #247: "dangling" references cause problems with cloning and transformations
Here is the example:

```
import pyomo.environ as pe
import idaes.core
import idaes.generic_models.unit_models
import idaes.generic_models.properties.swco2
from pyomo.network import Arc

m = pe.ConcreteModel()
m.fs = idaes.core.FlowsheetBlock()
m.fs.properties = idaes.generic_models.properties.swco2.SWCO2ParameterBlock()
m.fs.heater = idaes.generic_models.unit_models.Heater(default={'dynamic': False,
                                                               'property_package': m.fs.properties,
                                                               'has_pressure_change': True})
m.fs.heater2 = idaes.generic_models.unit_models.Heater(default={'dynamic': False,
                                                                'property_package': m.fs.properties,
                                                                'has_pressure_change': True})
m.fs.stream = Arc(source=m.fs.heater.outlet,
                  destination=m.fs.heater2.inlet)

try:
    m2 = pe.TransformationFactory('network.expand_arcs').create_using(m)
except Exception as err:
    print(err)
```

outputs

```
'NoneType' object has no attribute 'component'
```

Also, 

```
print(m.fs.heater.inlet.flow_mol[0] is m.fs.heater.control_volume.properties_in[0].flow_mol)  # True
```

outputs

```
True
```

but 

```
m2 = m.clone()
print(m2.fs.heater.inlet.flow_mol[0] is m2.fs.heater.control_volume.properties_in[0].flow_mol)  # False
```

outputs

```
False
```

Trying to clone `m2` produces the same error as above.
```
try:
    pe.TransformationFactory('network.expand_arcs').apply_to(m2)
except Exception as err:
    print(err)
```

outputs

```
'NoneType' object has no attribute 'component'
```

The transformation applied to the original model works because no cloning occurs.
```
pe.TransformationFactory('network.expand_arcs').apply_to(m) # okay
```

Note that some transformations do "work" (no exception is raised) after cloning. For example, the BigM transformation "works" with a cloned model. However, the result is that accessing anything on an inlet or outlet of a unit model in the resulting model does not do what is expected. You end up modifying the original model.

### PR DESCRIPTION
**Pull request by andrewlee94**
## Summary/Motivation:
Issue #247 identified issues with cloning models due to some `References` not being attached to actual model components. This PR addresses these dangling `References` and adds tests for cloning models.

## Changes proposed in this PR:
- Attached all dangling `References` to a `Block`
- Add tests for cloning models to ensure issue is resolved.

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Review by michaelbynum — DISMISSED** [review]
This looks perfect.

**Comment by michaelbynum** [comment]
According to the tests, this did not fix the problem. That really surprises me...

**Comment by andrewlee94** [comment]
@michaelbynum That is the problem - there is apparently something more involved here. I am hoping you or @jsiirola have some ideas on where to start.

**Comment by michaelbynum** [comment]
Sorry - I missed your comments on the issue. I'll look into it.

**Comment by Robbybp** [comment]
Added code looks good to me, but have no idea about the failing tests. Looks like `arc.expanded_block` is not getting set, but I quickly looked through `arc.py` and didn't see where it should be getting set.

**Comment by andrewlee94** [comment]
@michaelbynum I poked at this a bit more, and I can see some of what if happening. AS best I can tell the problem occurs in the `_apply_to` method of `ExpandArcs`.

Line 43 (calling `_add_blocks`) is supposed to create the expanded block, which it does successfully - I can display the `Block` at that point and see an empty `Block`.

However, once we get into the next part of the code, the `Arc` we get from the `Port` seems to be different from the `Arc` that was acted on in `_add_blocks`.

My best guess is that there is a disconnect between the `_add_blocks` method and the `rule` generated from the `Ports` (line 56), and that one of these is pointing to the original model, whilst the other is pointing to the cloned model. Hence, `add_blocks` is adding the expanded block to one instance, but the rule for adding the equality constraint is looking for this on the other instance.

**Comment by andrewlee94** [comment]
@michaelbynum @jsiirola I have confirmed my above suspicion. The `_add_blocks` method is acting on the original model, but the `rule` called for adding the equality constraints is trying to act on the cloned model.

**Comment by michaelbynum** [comment]
@andrewlee94 The Pyomo PR fixing the cloning of ports (pyomo/pyomo#1880) has been merged. Can we update this branch and rerun the tests?

**Comment by michaelbynum** [comment]
Can I just hit the "Update branch" button? I've never used that before...

**Comment by michaelbynum** [comment]
Ahh. Okay. Thanks.

**Comment by andrewlee94** [comment]
@jsiirola Will you update the IDAES Pyomo branch, or should I?

**Comment by andrewlee94** [comment]
@michaelbynum @jsiirola I fixed a conflict and updated the branch, so hopefully the tests pass now.

**Comment by michaelbynum** [comment]
Awesome, thanks!

**Comment by andrewlee94** [comment]
@michaelbynum I am looking into the test failure. A quick test on my local branch had these tests passing.

**Comment by andrewlee94** [comment]
@michaelbynum The failing tests can be ignored - they are checks using the last released version of Pyomo so they obviously fail and are not required.

I also added some extra asserts to make sure the model was being cloned and transformed as expected.

**Comment by andrewlee94** [comment]
@michaelbynum @jsiirola @Robbybp This is ready for review now. You can ignore the failing tests - we will discuss them at the dev call tomorrow and they are not required. (In brief - they are all tests running off the last release of Pyomo, which still has the bug).

**Review by michaelbynum — APPROVED** [review]
Looks great. Thanks @andrewlee94 !

File: idaes/core/unit_model.py
Comment (Robbybp) [inline idaes/core/unit_model.py:159]: Any reason `setattr` is used here instead of `blk.add_component`?

**Review by Robbybp — APPROVED** [review]
One question, but looks good.

File: idaes/core/unit_model.py
Comment (andrewlee94) [reply idaes/core/unit_model.py:159]: Only that it took a while for someone to teach me  about `add_component`, and I still forget it exists occasionally.

**Comment by lbianchi-lbl** [comment]
The new errors in the `Integration / Run examples` jobs seem to occur consistently on this branch. Some info:

- Only happens on Ubuntu, Windows works fine
- Four notebooks are affected, with the same (or very similar) error:

```txt
import pyomo.contrib.parmest.parmest as parmest
import numpy as np

parmest_results={}
# run parmest tool for each power bin
for i, group in df.groupby("bin_no"):
    pest = parmest.Estimator(get_model, list(group.index), theta_names, sse)
    obj, theta = pest.theta_est()
    print(f"Bin number: {i},  objective: {obj}")
    for k in theta:
        print(f"  {k}: {theta[k]}")
    parmest_results[i] = {'obj':obj, 'theta': theta}
------------------

---------------------------------------------------------------------------
DeferredImportError                       Traceback (most recent call last)
<ipython-input-1-bf089d8cc068> in <module>
      6 for i, group in df.groupby("bin_no"):
      7     pest = parmest.Estimator(get_model, list(group.index), theta_names, sse)
----> 8     obj, theta = pest.theta_est()
      9     print(f"Bin number: {i},  objective: {obj}")
     10     for k in theta:

/opt/hostedtoolcache/Python/3.8.8/x64/lib/python3.8/site-packages/pyomo/contrib/parmest/parmest.py in theta_est(self, solver, return_values, bootlist, calc_cov)
    835         assert isinstance(bootlist, (type(None), list))
    836 
--> 837         return self._Q_opt(solver=solver, return_values=return_values,
    838                            bootlist=bootlist, calc_cov=calc_cov)
    839 

/opt/hostedtoolcache/Python/3.8.8/x64/lib/python3.8/site-packages/pyomo/contrib/parmest/parmest.py in _Q_opt(self, ThetaVals, solver, return_values, bootlist, calc_cov)
    462         # (Bootstrap scenarios will use indirection through the bootlist)
    463         if bootlist is None:
--> 464             tree_model = _treemaker(self._numbers_list)
    465         else:
    466             tree_model = _treemaker(range(len(self._numbers_list)))

/opt/hostedtoolcache/Python/3.8.8/x64/lib/python3.8/site-packages/pyomo/contrib/parmest/parmest.py in _treemaker(scenlist)
    253 
    254     num_scenarios = len(scenlist)
--> 255     m = scenariotree.tree_structure_model.CreateAbstractScenarioTreeModel()
    256     m = m.create_instance()
    257     m.Stages.add('Stage1')

/opt/hostedtoolcache/Python/3.8.8/x64/lib/python3.8/site-packages/pyomo/common/dependencies.py in __getattr__(self, attr)
     85             for _sub in self._submodule_name[1:].split('.'):
     86                 _mod = getattr(_mod, _sub)
---> 87         return getattr(_mod, attr)
     88 
     89 

/opt/hostedtoolcache/Python/3.8.8/x64/lib/python3.8/site-packages/pyomo/common/dependencies.py in __getattr__(self, attr)
     45             raise AttributeError("'%s' object has no attribute '%s'"
     46                                  % (type(self).__name__, attr))
---> 47         raise DeferredImportError(self._error_message_)
     48 
     49     def generate_import_warning(self, logger='pyomo.common'):

DeferredImportError: The pysp.scenariotree module (an optional Pyomo dependency) failed to import
```

**Comment by andrewlee94** [comment]
@michaelbynum @jsiirola Any ideas about the test failures reported above?

**Comment by jsiirola** [comment]
I just added (as part of another PR) an extension to the `DeferredImportError` to include the original error that triggers the failed import that might help diagnose this error.  Can we test on Windows against Pyomo/main, or should we just bump the IDAES branch and rerun these tests?

**Comment by andrewlee94** [comment]
@jsiirola Interestingly all these failures were on Linux. However, I just replicated the failure locally so I can test it.

Do I just get the master branch of Pyomo?

**Comment by jsiirola** [comment]
Doh.  Missed the platform.  @andrewlee94, yes: just grab Pyomo/main.

**Comment by andrewlee94** [comment]
Output when running off Pyomo/main:

```
---------------------------------------------------------------------------
DeferredImportError                       Traceback (most recent call last)
<ipython-input-13-06556f9e1ec7> in <module>
      6 
      7 # Run parameter estimation using data subset 1
----> 8 obj_value_1, parameters_1 = pest_data_subset_1.theta_est()
      9 
     10 # Run parameter estimation using data subset 2

~/pyomo/pyomo/pyomo/contrib/parmest/parmest.py in theta_est(self, solver, return_values, bootlist, calc_cov)
    836 
    837         return self._Q_opt(solver=solver, return_values=return_values,
--> 838                            bootlist=bootlist, calc_cov=calc_cov)
    839 
    840 

~/pyomo/pyomo/pyomo/contrib/parmest/parmest.py in _Q_opt(self, ThetaVals, solver, return_values, bootlist, calc_cov)
    462         # (Bootstrap scenarios will use indirection through the bootlist)
    463         if bootlist is None:
--> 464             tree_model = _treemaker(self._numbers_list)
    465         else:
    466             tree_model = _treemaker(range(len(self._numbers_list)))

~/pyomo/pyomo/pyomo/contrib/parmest/parmest.py in _treemaker(scenlist)
    253 
    254     num_scenarios = len(scenlist)
--> 255     m = scenariotree.tree_structure_model.CreateAbstractScenarioTreeModel()
    256     m = m.create_instance()
    257     m.Stages.add('Stage1')

~/pyomo/pyomo/pyomo/common/dependencies.py in __getattr__(self, attr)
     85             for _sub in self._submodule_name[1:].split('.'):
     86                 _mod = getattr(_mod, _sub)
---> 87         return getattr(_mod, attr)
     88 
     89 

~/pyomo/pyomo/pyomo/common/dependencies.py in __getattr__(self, attr)
     45             raise AttributeError("'%s' object has no attribute '%s'"
     46                                  % (type(self).__name__, attr))
---> 47         raise DeferredImportError(self._error_message_)
     48 
     49     def generate_import_warning(self, logger='pyomo.common'):

DeferredImportError: The pysp.scenariotree module (an optional Pyomo dependency) failed to import: No module named 'pysp'
```

**Comment by Robbybp** [comment]
We ran into something similar in https://github.com/Pyomo/pyomo/pull/1613, and installing PySP did not seem to fix the problem, at least for me locally, as PySP depends on `pyomo.opt.parallel.pyro`, which no longer exists.

**Comment by lbianchi-lbl** [comment]
The latest run of `Integration / Run examples` seems to confirm what @Robbybp is saying: PySP gets installed correctly, but the tests are failing with the same `DeferredImportError` as before.

**Comment by jsiirola** [comment]
@lbianchi-lbl There was a PR to PySP that resolved that error (I just merged it).  There are still inexplicable test failures in PySP (we are still fighting that), but hopefully this will get us a lot closer to getting this PR working.

**Comment by lbianchi-lbl** [comment]
@jsiirola It looks like the fix worked! The `Integration / Run examples` jobs on Ubuntu are now passing.

**Comment by adowling2** [comment]
@lbianchi-lbl Here are the two outstanding errors:

```
  Collecting https://github.com/Pyomo/pysp/archive/master.zip
  Downloading https://github.com/Pyomo/pysp/archive/master.zip
  ERROR: Could not find a version that satisfies the requirement Pyomo>=6.0.0.dev0 (from pysp)
  ERROR: No matching distribution found for Pyomo>=6.0.0.dev0
  Error: Process completed with exit code 1.
```

```
  The following 5 notebooks had failures: 
    -  /home/runner/work/_temp/src/Examples/Advanced/DataRecon/econ_parmest_testing.ipynb
    -  /home/runner/work/_temp/src/Tutorials/Advanced/ParamEst/DMF_1_for_parameter_estimation_NRTL_using_unit_model_solution_testing.ipynb
    -  /home/runner/work/_temp/src/Tutorials/Advanced/ParamEst/DMF_2_flash_unit_Model_with_NRTL_solution_testing.ipynb
    -  /home/runner/work/_temp/src/Tutorials/Advanced/ParamEst/parameter_estimation_NRTL_using_state_block_solution_testing.ipynb
    -  /home/runner/work/_temp/src/Tutorials/Advanced/ParamEst/parameter_estimation_NRTL_using_unit_model_solution_testing.ipynb
  Notebook errors are in 'notebook-errors.txt'
  No parallel processing (number of workers = 1)
  Error: Process completed with exit code 1.
```

These notebooks all depend on ParmEst which uses PySP (but will not with Pyomo 6).

**Comment by lbianchi-lbl** [comment]
@adowling2 I'm not 100% sure but I think these are separate errors:

- The "Could not find a version that satisfies..." are probably the `Tests / pytest (rel/...)` jobs that use the latest Pyomo release (as opposed to the `IDAES` branch of 'IDAES/Pyomo`). As I understand these are expected to fail for this PR since the necessary changes to Pyomo do not appear in the latest Pyomo release
- The 5 notebooks that would fail on Ubuntu should be now fixed after @jsiirola merged in Pyomo/pysp#7
- We do have still some failures on the `Integration / Run examples` jobs (now, ironically, on Windows but not on Ubuntu), but the error looks different now:
```txt
...
 2021-04-09 03:18:57,321 INFO - [Worker 0] Convert notebook name=D:\a\_temp\src\Tutorials\Advanced\ParamEst\parameter_estimation_NRTL_using_unit_model_solution_testing.ipynb: begin
2021-04-09 03:18:57,321 INFO - [Worker 0] Running: parameter_estimation_NRTL_using_unit_model_solution_testing.ipynb
2021-04-09 03:18:57,452 INFO - [Worker 0] Stripped tags from: parameter_estimation_NRTL_using_unit_model_solution_testing.ipynb
2021-04-09 03:18:57,453 INFO - [Worker 0] Running notebook: parameter_estimation_NRTL_using_unit_model_solution.ipynb
2021-04-09 03:26:05,753 ERROR - [Worker 0] Failed due to error: 'ParallelNotebookWorker' object has no attribute 's'
2021-04-09 03:26:05,759 INFO - [Worker 0] Convert notebook name=D:\a\_temp\src\Tutorials\Advanced\ParamEst\parameter_estimation_NRTL_using_unit_model_solution_testing.ipynb: end, ok=False duration=428.4s
...
Run Jupyter notebooks
  Looking for notebooks in 'D:\a\_temp\src\Examples\Advanced\CustomUnitModels'
  Looking for notebooks in 'D:\a\_temp\src\Examples\Advanced\DataRecon'
  Looking for notebooks in 'D:\a\_temp\src\Examples\MatOpt'
  Looking for notebooks in 'D:\a\_temp\src\Examples\SurrMod\ALAMO'
  Looking for notebooks in 'D:\a\_temp\src\Examples\SurrMod\Helmet'
  Looking for notebooks in 'D:\a\_temp\src\Examples\SurrMod\PySMO'
  Looking for notebooks in 'D:\a\_temp\src\Examples\Tools'
  Looking for notebooks in 'D:\a\_temp\src\Examples\UnitModels'
  Looking for notebooks in 'D:\a\_temp\src\Tutorials\Advanced\ParamEst'
  Looking for notebooks in 'D:\a\_temp\src\Tutorials\Basics'
  Processed 10 directories in 1442.5s
      Out of 27 notebook(s), 26 converted and 1 failed
  The following 1 notebooks had failures: 
    -  D:\a\_temp\src\Tutorials\Advanced\ParamEst\parameter_estimation_NRTL_using_unit_model_solution_testing.ipynb
    Notebook errors are in 'notebook-errors.txt'
  No parallel processing (number of workers = 1)
```

**Comment by lbianchi-lbl** [comment]
Looking at the difference between the most recent `Integration / Run examples` runs on Windows and Ubuntu, it looks like on Windows the runtime is significantly larger: as far as I can tell, this affects notebooks in `Tutorials/Advanced/ParamEst`.

**Comment by andrewlee94** [comment]
@lbianchi-lbl It should be noted that that is one of the stochastic examples, so there is the possibility of variance between runs for that example. However, that should not be platform dependent so if we consistently see difference between platforms then we should look into the cause.

**Comment by lbianchi-lbl** [comment]
@andrewlee94 I've looked into the current error happening on Windows `AttributeError: 'ParallelNotebookWorker' has no attribute 's'`. I believe this is caused by a bug in the `build.py` script (from `examples-pse`) that occurs while catching a `TimeoutError` (https://github.com/IDAES/examples-pse/blob/722dd5037ca0c8087e03141cbc552edd75d1fd0b/build.py#L975-L977): the `AttributeError` then masks the error message that would describe the timeout.

I'm opening an issue in the examples directory addressing this (since it's something that would happen on any notebook that times out).

In the meantime, I'm thinking it could be useful to investigate why this timeout (which, presumably, would be caused by the affected notebooks taking longer to run) occurs on Windows with the versions of Pyomo and PySP used in this PR: @jsiirola, is this something expected, or that you all are aware of on the Pyomo side?

**Comment by jsiirola** [comment]
@lbianchi-lbl, I am not sure - apart from it is not unusual for Windows to be a *little* slower.  What is the timeout value (relative to the time that Linux usually takes)?  Also, I am not familiar with this code: what is `self.processor.preprocess` doing?

**Comment by lbianchi-lbl** [comment]
@jsiirola these are the runtimes for the `Tutorials/Advanced/ParamEst` notebooks:

Ubuntu:

```
2021-04-14 16:47:00,605 INFO - [Worker 0] Convert notebook name=/home/runner/work/_temp/src/Tutorials/Advanced/ParamEst/DMF_1_for_parameter_estimation_NRTL_using_unit_model_solution_testing.ipynb: end, ok=True duration=20.2s
2021-04-14 16:47:04,708 INFO - [Worker 0] Convert notebook name=/home/runner/work/_temp/src/Tutorials/Advanced/ParamEst/DMF_2_flash_unit_Model_with_NRTL_solution_testing.ipynb: end, ok=True duration=4.1s
2021-04-14 16:47:58,890 INFO - [Worker 0] Convert notebook name=/home/runner/work/_temp/src/Tutorials/Advanced/ParamEst/parameter_estimation_NRTL_using_state_block_solution_testing.ipynb: end, ok=True duration=54.2s
2021-04-14 16:47:58,890 INFO - [Worker 0] Convert notebook name=/home/runner/work/_temp/src/Tutorials/Advanced/ParamEst/parameter_estimation_NRTL_using_unit_model_solution_testing.ipynb: begin
2021-04-14 16:49:28,078 INFO - [Worker 0] Convert notebook name=/home/runner/work/_temp/src/Tutorials/Advanced/ParamEst/parameter_estimation_NRTL_using_unit_model_solution_testing.ipynb: end, ok=True duration=89.2s
```

Windows:

```
2021-04-14 16:53:51,017 INFO - [Worker 0] Convert notebook name=D:\a\_temp\src\Tutorials\Advanced\ParamEst\DMF_1_for_parameter_estimation_NRTL_using_unit_model_solution_testing.ipynb: end, ok=True duration=127.3s
2021-04-14 16:53:58,159 INFO - [Worker 0] Convert notebook name=D:\a\_temp\src\Tutorials\Advanced\ParamEst\DMF_2_flash_unit_Model_with_NRTL_solution_testing.ipynb: end, ok=True duration=7.1s
2021-04-14 16:59:02,981 INFO - [Worker 0] Convert notebook name=D:\a\_temp\src\Tutorials\Advanced\ParamEst\parameter_estimation_NRTL_using_state_block_solution_testing.ipynb: end, ok=True duration=304.8s
2021-04-14 16:59:02,991 INFO - [Worker 0] Convert notebook name=D:\a\_temp\src\Tutorials\Advanced\ParamEst\parameter_estimation_NRTL_using_unit_model_solution_testing.ipynb: begin
2021-04-14 17:06:11,385 ERROR - [Worker 0] Failed due to error: 'ParallelNotebookWorker' object has no attribute 's'
2021-04-14 17:06:11,386 INFO - [Worker 0] Convert notebook name=D:\a\_temp\src\Tutorials\Advanced\ParamEst\parameter_estimation_NRTL_using_unit_model_solution_testing.ipynb: end, ok=False duration=428.4s
```

As for the question of what `self.processor.preprocess()` is doing: it's a bit of a rabbit hole, but from my understanding it's basically a few layers of wrappers around `nbconvert.NotebookClient.execute()`, which runs a notebook cell by cell. I believe that the value for `timeout` that ultimately affects when the `TimeoutError` is thrown is set [here](https://github.com/IDAES/examples-pse/blob/722dd5037ca0c8087e03141cbc552edd75d1fd0b/build.py#L455), but again, without having the full stack trace (and more knowledge of the code in `build.py`) this is just a semi-educated guess.

**Comment by andrewlee94** [comment]
@lbianchi-lbl I also assume there is no easy way of determining where all the run time is being spent in these test? I.e. how much is in converting the notebook, how much is Pyomo build time and how much is spent in IPOPT.

**Comment by lbianchi-lbl** [comment]
@andrewlee94 that's a good point. I'd say that one reasonable thing to try could be taking the whole notebook machinery out of the equation and see if the problem persists. I can try to simply copy and paste the code in the offending notebooks in a `.py` file and check the runtime using that.

The next step could be trying to run it through the default Python profiler: this likely won't give us too much (if any) information about anything that runs in another process, but could still be useful as a quick check.

I'll try to put together something this afternoon and report back later.

**Comment by lbianchi-lbl** [comment]
I've looked into one of the `ParamEst` notebooks, `parameter_estimation_NRTL_using_state_block_solution_testing`, in more detail. Here's a summary of what I tried:

- I extracted the Python code in the notebook to a standalone file (`parmest_state_block.py`, attached inside `parmest.zip`)
- Running the file as a script with `python parmest_state_block.py` on Ubuntu and Windows seems to confirm the fact that the runtime on Windows (~400 s) is much longer than on Ubuntu (~40 s)
  - My testing on Windows was on a VM
  - Comparing the outputs (`ubuntu.log` and `windows.log`, attached), my untrained eye didn't spot any glaring differences between the two
  - The output of `pip list` for each system is attached
- Running the file through the standard `cProfile` compiler (`python -m cProfile -o parmest.prof parmest_state_block.py`) on the two OSes resulted in the two `.prof` files attached inside `parmest.zip`
  - These can be visualized interactively with the [`snakeviz`](https://jiffyclub.github.io/snakeviz/) tool (`pip install snakeviz && snakeviz parmest.prof`)
  - The information on Windows looks partial or incomplete (possibly because of differences in how the two OSes collect metrics from subprocesses?)
- Finally, I ran the notebook manually on Windows, with similarly long runtimes. A PDF export of the finished output is attached (HTML version inside `parmest.zip`)
  - To be able to run the notebook, I had to apply the fix needed for Jupyter on Python 3.8 documented in #258

[parameter_estimation_NRTL_using_state_block_solution_testing.pdf](https://github.com/IDAES/idaes-pse/files/6314195/parameter_estimation_NRTL_using_state_block_solution_testing.pdf)
[parmest.zip](https://github.com/IDAES/idaes-pse/files/6314196/parmest.zip)
[pip-list-ubuntu.txt](https://github.com/IDAES/idaes-pse/files/6314197/pip-list-ubuntu.txt)
[pip-list-win.txt](https://github.com/IDAES/idaes-pse/files/6314198/pip-list-win.txt)
[ubuntu.log](https://github.com/IDAES/idaes-pse/files/6314199/ubuntu.log)
[win.log](https://github.com/IDAES/idaes-pse/files/6314200/win.log)

**Comment by lbianchi-lbl** [comment]
@andrewlee94 @jsiirola I've tried increasing the `timeout` parameter in the examples build configuration used by `Integration / Run examples` from 300 to 600: the Windows errors are gone, and all integration tests now pass. In the narrowest sense, the remaining issue preventing all checks that are expected to pass to succeed is now fixed.

More broadly, I imagine that the question is how to deal with the apparent parameter estimation performance hit on Windows:

- Should we merge this PR in, and then investigate the parest Windows performance issue in a separate PR?
- Should we refrain from merging in this PR until the cause is known (and possibly fixed)?
- Should this be tackled from the Pyomo side, from the IDAES side, or both?

(This is just a quick list of things I was able to come up with now, so it's definitely not meant to be complete or accurate)

**Comment by adowling2** [comment]
@lbianchi-lbl This is holding up PR #131 which we would like to merge in 8 days (dev call next week). My preference is to investigate the parmest performance in Windows on a separate PR.

**Comment by adowling2** [comment]
@lbianchi-lbl In the notebook PDF you posed, I see Ipopt maxed out at 3000 iterations for one of the solves. If is possible for you to generate the same notebook PDF on Ubuntu? I'm curious if that is happening on both test systems. This might be a dead end through, as Ipopt only takes ~14 seconds in that solve that goes to the maximum number of iterations.

If I add up the Ipopt solve times reported in the notebook PDF, I get ~20 seconds. This is consistent with the run time of ~40 seconds on Ubuntu. This suggests the issue is either with building the Pyomo model or something else; it is not a solver problem.

**Comment by lbianchi-lbl** [comment]
@adowling2 this makes sense. I myself don't have particular preferences: I just wanted to raise this question so that everyone is aware of the existing issues when evaluating pros and cons of this decision.

For my side, following last week's discussions, I went ahead and removed the `pytest-rel` checks (that are expected to fail following the changes introduced in this PR) from the `Tests` workflow that runs on PRs and on `main`: as far as I'm concerned, we should now have a (figurative and literal) green light for this PR.

**Comment by andrewlee94** [comment]
@lbianchi-lbl The `parmest` issue is separate from this PR, and is affecting all other PRs as well, so I think we should also go ahead with the patch fix and investigate the `parmest` issue in a separate PR.

**Review by Andresj89 — APPROVED** [review]
LGTM

**Review by michaelbynum — APPROVED** [review]
Looks good!

**Pull request merged**
